File: kite/main/control/actuator.c
#include "../../../common/helpers/math.h"
#include "actuator.h"

void initActuator(Actuator* actuator, float speed, float minValue, float maxValue){
	
	actuator->targetValue = minValue;
	actuator->speed = speed;
	actuator->minValue = minValue;
	actuator->maxValue = maxValue;
	actuator->currentValue = (minValue+maxValue)*0.5;
}

void stepActuator(Actuator* actuator, float time_difference){
	
	if(actuator->targetValue > actuator->currentValue){
		actuator->currentValue += time_difference * actuator->speed;
		actuator->currentValue = clamp(actuator->currentValue, actuator->minValue, actuator->targetValue);// avoid shooting over targetValue
	}else{
		actuator->currentValue -= time_difference * actuator->speed;
		actuator->currentValue = clamp(actuator->currentValue, actuator->targetValue, actuator->maxValue);// avoid shooting under targetValue
	}
	actuator->currentValue = clamp(actuator->currentValue, actuator->minValue, actuator->maxValue);
	
}

float getValueActuator(Actuator* actuator){
	
	return actuator->currentValue;
	
}

void setTargetValueActuator(Actuator* actuator, float value){
	
	actuator->targetValue = value;
}

File: simulation/Actuator.js
"use strict";

class Actuator{
	
	constructor(speed, minValue, maxValue){
		this.targetValue = 0;
		this.speed = speed;
		this.minValue = minValue;
		this.maxValue = maxValue;
		this.currentValue = 0;
	}
	
	step(time_difference){
		if(this.targetValue > this.currentValue){
			this.currentValue += time_difference * this.speed;
			this.currentValue = clamp(this.currentValue, this.minValue, this.targetValue);
		}else{
			this.currentValue -= time_difference * this.speed;
			this.currentValue = clamp(this.currentValue, this.targetValue, this.maxValue);
		}
		this.currentValue = clamp(this.currentValue, this.minValue, this.maxValue);
	}
	
	getValue(){
		return this.currentValue;
	}
	
	setTargetValue(tV){
		this.targetValue = tV;
	}
}
